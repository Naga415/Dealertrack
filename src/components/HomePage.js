"use strict";
import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { bindActionCreators } from "redux";

import * as FeesActions from "../actions/FeesActions";

import FullPageModal from "@coxautokc/fusion-ui-components/lib/FullPageModal";
import ModalFooter from "./ModalFooter/ModalFooter";
import FeesFormHeader from "./FeesFormHeader/FeesFormHeader";
import FeesFormElements from "./FeesFormElements/FeesFormElements";
import "./HomePage.scss";
import "font-awesome/css/font-awesome.css";

export class HomePage extends Component {
  componentDidMount() {
    this.props.fetchExistingFeeQuoteFee(this.props.dealId, this.props.scenarioId);
  }

  cancel = () => {
    this.props.closeDialog();
    this.props.cancelEditing();
  };

  save = () => {
    const dataToSend = this.filterQuoteFees(this.props.quoteFees);
    const actionObject = this.props.validateQuoteFees(dataToSend);

    if (!actionObject.payload.formHasError) {
      const totalFees = dataToSend.reduce((accu, data) => {
        return accu + data["feeAmount"];
      }, 0);
      this.props.saveQuoteFees();
      this.props.pushFeeQuoteFee(dataToSend);
      this.props.populateValue(this.props.name, totalFees.toString());
      this.props.closeDialog();
    }
  };

  filterQuoteFees = quoteFees => {
    const dataToSend = quoteFees
      .filter(quoteFee => {
        return (
          quoteFee.feeType.length ||
          quoteFee.feeCustomName.length ||
          quoteFee.feeAmount.length ||
          quoteFee.paidTo.length
        );
      })
      .map(quoteFee => {
        // remove coma from numbers
        return Object.assign({}, quoteFee, {
          feeAmount: Number(quoteFee["feeAmount"].replace(/,/g, "")),
          zipCode: quoteFee["zipCode"] || this.props.zipCode
        });
      });

    return dataToSend;
  };

  showError = () => {
    const errorToShow = Object.keys(this.props.error).filter(key => {
      return this.props.error[key];
    });

    return "Please Select Fields: " + errorToShow.join(", ") + ".";
  };

  render() {
    const { closeDialog, dialogStatus } = this.props;
    return (
      <FullPageModal
        htmlId="Fees-Modal"
        className="sample-fullpage-modal"
        show={dialogStatus}
        header={<FullPageModal.Title>EDIT FEES</FullPageModal.Title>}
        onHide={closeDialog}
        footer={
          <ModalFooter
            displayResetBtn={true}
            modalName="fees"
            save={this.save}
            cancel={this.cancel}
            resetData={this.props.refreshForm}
          />
        }
      >
        <div>
          {this.props.formHasError && (
            <div className="alert alert-danger">{this.showError()}</div>
          )}
          <FeesFormHeader />
          <FeesFormElements
            grouppedFees={this.props.grouppedFees}
            handleChangeFromForm={this.props.handleChangeFromForm}
            handleChangeFromInput={this.props.handleChangeFromInput}
            handleBlurFromInput={this.props.handleBlurFromInput}
            appendFee={this.props.appendFee}
            deleteFee={this.props.deleteFee}
            quoteFees={this.props.quoteFees}
          />
          <p>
            <b>*Lender Fees</b> - The fee amount from lender programs will be
            used
          </p>
        </div>
      </FullPageModal>
    );
  }
}

HomePage.propTypes = {
  closeDialog: PropTypes.func.isRequired,
  dialogStatus: PropTypes.bool.isRequired,
  name: PropTypes.string.isRequired,
  populateValue: PropTypes.func.isRequired,
  fetchExistingFeeQuoteFee: PropTypes.func,
  dealId: PropTypes.number.isRequired,
  scenarioId: PropTypes.number.isRequired,
  cancelEditing: PropTypes.func,
  quoteFees: PropTypes.func,
  validateQuoteFees: PropTypes.func,
  saveQuoteFees: PropTypes.func,
  pushFeeQuoteFee: PropTypes.func,
  zipCode: PropTypes.number.isRequired,
  error: PropTypes.object,
  refreshForm: PropTypes.func,
  formHasError: PropTypes.bool,
  grouppedFees: PropTypes.object,
  handleChangeFromForm: PropTypes.func,
  handleChangeFromInput: PropTypes.func,
  handleBlurFromInput: PropTypes.func,
  appendFee: PropTypes.func,
  deleteFee: PropTypes.func,
  quoteFees: PropTypes.array
};

function mapStateToProps(state) {
  return {
    formHasError: state.feesReducer.formHasError,
    error: state.feesReducer.error,
    fees: state.feesReducer.fees,
    grouppedFees: state.feesReducer.grouppedFees,
    quoteFees: state.feesReducer.quoteFees,
    savedQuoteFees: state.feesReducer.savedQuoteFees
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(FeesActions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
