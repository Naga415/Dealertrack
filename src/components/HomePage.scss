@import "~@coxautokc/fusion-theme/dist/fusion-variables.scss";

#Fees-Modal {
  h4 {
    font-size: 16px;
    font-weight: bold;
  }

  .modal-dialog {
    width: 50%;
    height: 50%;
    margin: 0;
    padding: 0;
    position: absolute;
    top: 12.5%;
    right: 0;
    bottom: 0;
    left: 25%;
  }

  .modal-body {
    padding-left: 3%;
    padding-right: 3%;
  }
}
