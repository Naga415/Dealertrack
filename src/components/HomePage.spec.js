"use strict";
import React from "react";
import sinon from "sinon";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { HomePage } from "./HomePage";

const closeDialog = sinon.spy();
const populateValue = sinon.spy();
const deleteFee = sinon.spy();
const handleChangeFromForm = sinon.spy();
const handleBlurFromInput = sinon.spy();
const handleChangeFromInput = sinon.spy();
const appendFee = sinon.spy();
const fetchExistingFeeQuoteFee = sinon.stub();

afterEach(() => {
  closeDialog.reset();
  populateValue.reset();
  deleteFee.reset();
  handleChangeFromForm.reset();
  handleBlurFromInput.reset();
  handleChangeFromInput.reset();
  appendFee.reset();
});

describe("HomePage specs", () => {
  const props = {
    closeDialog,
    populateValue,
    deleteFee,
    handleChangeFromForm,
    handleBlurFromInput,
    handleChangeFromInput,
    appendFee,
    fetchExistingFeeQuoteFee,
    quoteFees: [],
    grouppedFees: {},
    name: "fees",
    dialogStatus: false,
    dealId: 123,
    scenarioId: 234,
    zipCode: 11111
  };

  it("renders <HomePage />", () => {
    const component = renderer.create(<HomePage {...props} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("filterQuoteFees returns formatted array", () => {
    const component = shallow(<HomePage {...props} />);
    const qf = [
      {
        feeId: 0,
        feeCategory: "",
        feeCategoryId: 0,
        feeType: "feeType-0",
        feeTypeId: 0,
        state: "",
        feeCustomName: "feeCustomName-0",
        capitalizedIndicator: false,
        feeAmount: "12.00",
        paidTo: "paidTo-0"
      },
      {
        feeId: 1,
        feeCategory: "",
        feeCategoryId: 0,
        feeType: "feeType-1",
        feeTypeId: 0,
        state: "",
        feeCustomName: "feeCustomName-1",
        capitalizedIndicator: true,
        feeAmount: "1,333.00",
        paidTo: "paidTo-1"
      },
      {
        feeId: 0,
        feeCategory: "",
        feeCategoryId: 0,
        feeType: "",
        feeTypeId: 0,
        state: "",
        feeCustomName: "",
        capitalizedIndicator: false,
        feeAmount: "",
        paidTo: ""
      }
    ];

    const returnedValue = component.instance().filterQuoteFees(qf);

    expect(returnedValue.length).toEqual(2);

    expect(returnedValue[0].feeAmount).toEqual(12);
    expect(returnedValue[0].capitalizedIndicator).toBeFalsy();
    expect(returnedValue[0].zipCode).toEqual(11111);

    expect(returnedValue[1].feeAmount).toEqual(1333);
    expect(returnedValue[1].capitalizedIndicator).toBeTruthy();
    expect(returnedValue[0].zipCode).toEqual(11111);
  });

  it("showError returns correct error msg", () => {
    const tmpProps = Object.assign({}, props, {
      error: {
        "Fee Type": false,
        "Fee Name": true,
        Amount: false,
        "Paid To Code": true
      }
    });
    const component = shallow(<HomePage {...tmpProps} />);
    const returnedMSG = component.instance().showError();

    expect(returnedMSG).toEqual(
      "Please Select Fields: Fee Name, Paid To Code."
    );
  });
});
