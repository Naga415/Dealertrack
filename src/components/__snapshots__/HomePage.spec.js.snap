// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage specs renders <HomePage /> 1`] = `<div />`;
